### DISCUSSION TITLE
Crazyflie 2.1+ falls out of air (android app control)

### DISCUSSION THREAD

**Discussion by cameronalred** [bitcraze/discussions#1710] [Q&A]
This month our lab purchased a set of 3 crazyflie 2.1+s. We forgot to get a crazyradio for PC control and will order a few at some point, but in the meantime I am test flying them with the Android app. I've been having the same issue with *all 3* Crazyflies:

-I plug in to charge and wait for the blue LED to stop flashing
-I connect to the app
-I am able to take off and fly for some random time, sometimes up to a minute, sometime as little as 9 seconds
-The quadcopter suddenly loses all power and drops to the floor, before restarting itself post-crash
-I have to power cycle it to be able to get it to connect and fly again.

It's worth noting that the console messages I get appear cut off, such as:

unning!
3.11) CLEAN
ash!
ce.
K].
OK]
]
(1) estimator
roller
river
tem.c.
21)
]
]

 I am using Android 13 on a Samsung Galaxy S20 Ultra.
 
 I am hoping that this issue is at least specific to the app control, and doesn't persist with crazyradio control, but I wanted to bring it up in case it's indicating a deeper problem.

**Comment by zanecodes** [comment]
I'm also experiencing this with a brand new Crazyflie 2.1+, on a Pixel 6a running Android 15. I don't have a CrazyRadio, I only have this Android device and a Surface Pro 7, which does have Bluetooth LE support, but cfclient still doesn't seem to have BLE support, and I was unable to get the [Windows UAP client](https://github.com/bitcraze/crazyflie2-windows-uap-client/tree/master) to run after compiling it with Visual Studio 2022 Community Edition, so I'm not sure if I'm able to flash a new firmware.

**Comment by ToveRumar** [comment]
Hi!

Im sorry you are experiencing these issues. 

Just as you say you need a Crazyradio to communicate from your PC. 
Using the app to control the drone requires practice to have the drone flying stably for a longer time. 

Im not sure exactly what is happening with your drone when it is crashing. A video would be very helpful. 
It sounds like the battery might be very close to empty. This will cause the exact behavior you explain. Does this happen when the battery is fully charged?

**Reply by zanecodes** [reply]
I am experiencing this issue consistently with a freshly fully charged Crazyflie. Just like @cameronalred I am only able to fly it for anywhere from 10 to 90 seconds or so before the drone abruptly loses power, falls from the air, and then reboots itself without any intervention from me. Then it blinks the "self-test failed" pattern on the red LED until I manually power cycle it.

When this happens, I also find that I am unable to reconnect over Bluetooth LE on my Android device (exhibiting a BLE connection timeout error) until I also disable and re-enable Bluetooth on my Pixel 6a.

**Reply by cameronalred** [reply]
Battery issues are the first thing I suspected, but as I stated in my post I plug it in and wait for the blue LED to go solid (I'm assuming this means it is fully charged?) and the issue still occurs. It's also happening equally to all three crazyflies, so it seems like a software issue rather than a hardware one. Since I'm using my phone for piloting a video might be tricky to get; I can see if a coworker can film it for me, but my best description is that one moment I'm achieving a fairly stable hover a few feet above the floor and without warning the motors completely cut to 0% and it falls straight down to the floor - it's not a gradual loss of power or at all, nor is it a pilot-induced crash.

**Reply by ToveRumar** [reply]
Thank you for the video, it was very helpful! 
I have tried reproducing this issue but I can not seem to do be able to. 

Im suspecting a battery issue, or maybe some other hardware issue.  It does not look like the battery discharges, just like you say. But maybe there is some issue with the battery not being able to give enough power or something similar. Are you using the same battery for all three drones? Are the drones and battery new?

Do you know what FW is running in the drones? 
When it is crashed, can you connect it with usb to your pc to be able to look at the console logs? It might be possible even though you can not connect with BLE. Sometimes connection "fails" with USB when the self test light goes on, but its still connected enough to see the console logs. This will give us info about what might be going on.

**Reply by zanecodes** [reply]
I'm not sure if you're referring to my video, or if @cameronalred sent a separate video somewhere besides GitHub (I am not affiliated with @cameronalred, I'm just experiencing the same problem).

My drone and battery are both brand new out of the box as of a few weeks ago.

My Crazyflie came with `2023.11` out of the box. Today I installed `cfclient` on my PC and connected the Crazyflie over micro-USB to see if I could flash it in DFU mode to the latest firmware. After a few missteps, my Crazyflie is now on firmware `2024.10.2`, but it continues to have the same issue, randomly rebooting after a minute or two of flying. Luckily, I can now plug it into my PC and see the entire untruncated logs in `cfclient`, where I can see the following assertion failure immediately after the reboot occurs:
```
SYS: Assert failed at ../src/drivers/src/uart_syslink.c:549
```

I also ran the battery and propellor tests in `cfclient` both before and after updating the firmware. Both tests passed on `2023.11`, and the propellor test failed on `2024.10.2` with [this error](https://github.com/bitcraze/crazyflie-firmware/issues/1430).

**Comment by zanecodes** [comment]
https://github.com/user-attachments/assets/a44c1771-aa4f-497c-bbec-4dd8e4df137c

Here is a video demonstrating the issue. This was about 2 minutes after after unplugging the Crazyflie from a charger, after the blue light stopped blinking and stayed solid. I was able to make two short hops before getting this clip but flew out of the camera's view each time, as shown by the `SUP:` messages in the console.

**Comment by ToveRumar** [comment]
Hi!

After some further testing we have found that this is an issue when using  at least some android versions. 

We will investigate this issue more to see what is going wrong. I have created an issue ticket on the matter; https://github.com/bitcraze/crazyflie-firmware/issues/1447 

You can probably mitigate it by increasing the queue size or dropping the packet if the queue is full (by not asserting). However it is not a full solution. 

I would anyways recommend to get a Crazyradio since this is the preferred way to fly it. 

Thank you for finding this bug!

**Reply by zanecodes** [reply]
Excellent, thanks for investigating! I came to roughly the same conclusion after taking a look at the firmware source. I think the truncated console logs may possibly be related, and another clue. I may try to see if I can set up a local development environment to build the firmware using [Devbox](https://www.jetify.com/docs/devbox/) or plain [Nix](https://nixos.org/), to avoid having to set up the Crazyflie development VM.

I don't currently have a Crazyradio, but in the meantime I'm taking a look at the possibility of developing a web Crazyflie client using the [Web Bluetooth](https://developer.mozilla.org/en-US/docs/Web/API/Web_Bluetooth_API) and [Gamepad](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API) APIs, as well as possibly facilitating firmware updates via a web app.

**Reply by ToveRumar** [reply]
Thank you for your help @zanecodes, this is very valuable!

Ill close this discussion now, and this issue can be followed here; https://github.com/bitcraze/crazyflie-firmware/issues/1447
